### PR TITLE
Allow empty <verbatim></verbatim> tag

### PR DESCRIPTION
**Pull request by labdsf**
Behavior is made consistent with Emacs Muse.

### CODE REVIEW

File: t/testfiles/verb.exp.ltx [hunk 1]
```diff
@@ -13,7 +16,7 @@ Test *M \textbackslash{}\{\#!\_<"> e*\forcelinebreak
 Test **M \textbackslash{}\{\#!\_<"> e**\forcelinebreak 
 Test ***M \textbackslash{}\{\#!\_<"> e***\forcelinebreak 
 Test =M \textbackslash{}\{\#!\_<"> e=\forcelinebreak 
-Test <verbatim> </verbatim>
+Test <verbatim> <\Slash{}verbatim>
```
Comment (labdsf) [inline t/testfiles/verb.exp.ltx]: Not sure what happened here, but it does not seem harmful

Comment (melmothx) [reply t/testfiles/verb.exp.ltx]: It comes out from this sequence `Test <verbatim><verbatim></verbatim> <verbatim></verbatim></verbatim>`
 . Your output looks more correct, but I believe only by chance. Needs further digging an more tests. In the exiting code the characters inside verbatim are just escaped, so the slash is kept as is. Not a big deal, as it's not a special character per se. However, the testfile change points out that the sequence is interpreted differently (because the second escaping is considered empty and the last </verbatim> is left as is.

It would be good if you added more tests with corner cases like this.

**Comment by melmothx** [comment]
I honestly don't recall how Emacs muse behave with `<em></em>` tags. This implementation asks for material inside, because it's just noise. Machine doesn't care, but the person reading it does. So I don't like to have source files with stuff like `<em></em><em></em><em></em><em></em><em></em>` which is just noise. It would be good if we reach an agreement here how to behave. (btw, Emacs muse doesn't have em/strong inline tags).

**Comment by labdsf** [comment]
I completely agree that empty tags is noise, thus should not be produced. But if someone accidentally produces noisy input, consumer should not choke on it. That is in accordance with Postel's law ("be generous in what you accept and strict in what you produce"). If you want to prevent users from producing bad text it might make sense to integrate some sort of "lint" tool to warn about possible improvements. Postel's law may be controversial, but opposite approach (fail fast) is not suitable for human-writable format. Certainly we don't want parser to fail on trailing whitespace, which is, likewise, a noise.

In pandoc I currently try to make a round-trip test based on QuickCheck. It generates random documents as abstract syntax trees. I write them into Muse format and read back to remove elements which are impossible to represent in text. Then I write and read it again to compare with result of the first pass and report error if it does not. If I manage to make it work once, this test will prevent Muse writer from producing text that is not readable by reader in the future.

Generated abstract syntax trees are not necessarily normal, they may contain elements like empty paragraphs, adjacent emphasis etc. Normalizing them before starting a test would not help. For example, AST `[Para [Code ("",[],[]) "</code>"]]`, represented in Muse, is `<code><</code><code>/code></code>`. Read it back and you get adjacent code elements. Surely in this case it can be represented as `=</code>=`, but more complex example can be easily produced to fail whatever heuristic I use to choose between `=` and `<code>`.

Therefore, I should put normalization somewhere in the loop while trying to keep implementation as simple as possible. So I want to put it either in the reader or writer, but not both. One approach is to put it in the reader and let writer output AST as it is. This approach fails because other readers or filters (which can be written in Lua by non-technical users in adhoc manner) may not normalize ASTs they produce, so converting from some other format to Muse may result in ugly or incorrect output. Existing markdown writer fails easily with AST `[Para [Emph [Str "abc"], Emph [Str "abc"]]]` as input. It produces `*abc**abc**` which is not correct markdown, instead of `*abc\*\*abc*`. That is why I want to put normalization into writer. I will normalize whatever I want to write by joining `Emph` in AST, escape whatever strings it contains and output it in the simplest way possible. This will ensure output is at least correct. Still, by constructing complicated examples it will be possible to make it produce things like `<code><</code><code>/code></code>`, so reader should be prepared to read such things.

Trying to make this QuickCheck test work is hard enough even without additional restrictions on output. Code for similar markdown test is currently commented for markdown because developers failed to make it work. It is not easy, but will greatly simplify maintenance. Even trying to make it work reveals many bugs in both reader and writer.

That is why I try to remove unnecessary incompatibilities where Text::Amuse introduces them, like not accepting empty tags, not accepting `<comment>` inline etc. Such not normal elements are needed in some core cases like separating two nested lists.

Once QuickCheck test is in place, I will be able to work on optimizing output, like using `*` where possible to make output look better and keep compatibility with Emacs Muse. Without such a test I will certainly miss some corner cases and produce incorrect output.

**Comment by melmothx** [comment]
Ok, I get the problem. From the work you're doing on the pandoc side,
this implementation has much to gain for sure.

So, it's not only the verbatim which needs to accepted as empty tag, but
also the other inline tags.

Regarding the inline `<comment>`, please open another ticket (even if I
don't like much the idea of an inlined comment, but hey).

-- 
Marco

**Comment by melmothx** [comment]
This has been merged in adf8326 but amended to be general for all the inline tags. They are accepted but removed from output.
